library(data.table)
library(dplyr)
library(ggplot2)
library(forcats) #for fct_inorder
library(gridExtra)
library(ape)
library(phytools)
library(stringr)
library(readxl)
library(stringi)
library(RColorBrewer)
library(ggpubr)
library(cowplot)

setwd("project_Listeria/")

################################
########### Inputs ###########
################################

#Anootation of genes with thier location in MGE (generated by mergeAnnotation.R)
annot <- fread("harmionizedAnnotation_correctGeneNames.tsv")
annotCare =  annot[gene!="blaZ",] #reviewer does not want blaZ

##Metadata
meta <- data.table(read_excel("SupplementaryTables.xlsx", "Table_S1", skip=1))


#Get the tree
tree <- read.tree("core_gene_alignment_snp_GTR-R_rooted.contree.nwk")

#Added outgroup for rooting, but delete for figure because branch is too long
tree <- drop.tip(tree, "Linnocua")

hc <- as.hclust(force.ultrametric(tree))
dend <- as.dendrogram(hc)

################################
# Set colors for lineages and CC 
################################

#Too many CC to give colors to all of them. I only give color to those that have at least 30 assemblie in the dataset
#The other CC all get the color gray  "#CCCCCC"
colCC_tb <- data.table(colCC = brewer.pal(n = 10, name = 'Spectral'), CC = names(table(meta$clonal_complex)[table(meta$clonal_complex)>=30]))
meta[, colCC := colCC_tb[chmatch(meta$clonal_complex, CC), colCC]]
meta[is.na(colCC), ]$colCC = "#CCCCCC"
annotCare[, CC := meta[chmatch(annotCare$assembly, ID), clonal_complex]]
annotCare[, colCC := meta[chmatch(annotCare$assembly, ID), colCC]]

#Color for lineages
colLineage <- distinct(annotCare[, c("assembly", "phylogenetic_lineage", "CC", "colCC")]) %>% 
  mutate(., colLineage = 
           ifelse(phylogenetic_lineage=="I", "#56B4E9",
                  ifelse(phylogenetic_lineage=="II", "#009E73", 
                         ifelse(phylogenetic_lineage=="III", "#F0E442",
                                "#CC79A7"))))

colLineage <- colLineage[match(tree$tip.label, assembly),]

#FOr the figure, to see when not the same CC, alternate between two shade of gray, instead of the same gray for everyone
colLineage$colCC2 = colLineage$colCC
grays <- c("#CCCCCC", "#636161")
prev = colLineage[1, c("CC", 'colCC2')]
for(i in 2:nrow(colLineage)){
  nex <- colLineage[i, c("CC", 'colCC2')]
  if(prev$colCC %in% grays & nex$colCC %in% grays & prev$CC==nex$CC){
    colLineage[i, ]$colCC2 = prev$colCC2
  } else if(prev$colCC %in% grays & nex$colCC %in% grays & prev$CC!=nex$CC){
    colLineage[i, ]$colCC2 = grays[!grays %in% prev$colCC2]
  }
  prev = colLineage[i, c("CC", 'colCC2')]
}

#Save the colors 
fwrite(colLineage, "colLineage.tsv", sep = "\t") #


##########################
######### Matrix #########
##########################


#Read annotation all MGE we retained in the final analysis
allMGE <- fread("all_936SRR_MGE_filtered_nodes.tbl")

### Start with PLASMIDS's matrix

#Create the empty matrix
plasmids <- allMGE[type=="plasmid"] %>%
  mutate(., primary_cluster_id = splitToColumns(nodeName, split = ":", columns = 2)) %>%
  mutate(., contig = sub(".*_", "", splitToColumns(MGEname, split = ":", columns = 2))) %>%
  mutate(., loc = paste(assembly, contig, sep = ":"))


plasmidMatrix <- matrix(nrow = 936, ncol = length(unique(plasmids$primary_cluster_id)))
rownames(plasmidMatrix) <- tree$tip.label
colnames(plasmidMatrix) <- sort(unique(plasmids$primary_cluster_id))


#Fill the matrix
for(j in colnames(plasmidMatrix)){ 
  for(i in rownames(plasmidMatrix)){ 
    print(i)
    toFill <-  plasmids[assembly == i & primary_cluster_id==j,]
    
    if(nrow(toFill)>0){
      #give to plasmid the same value as for genes in plasmid
      plasmidMatrix[i,j] =  2
    } else {
      plasmidMatrix[i,j] =  0
    }
  } 
} 

###########

### 2nd matrix = TRANSPOSON

#allMGE do not contain location transposons. So instead use annotation from bacant
bacant <- fread("all_annotation_named_DBv3.1.tsv", 
                col.names=c("contig", "element", "element2", "start", "end", "V6", "strand", "V8", "tag", "assembly"))  
transposons <- bacant[element=="transposon"]
transposons$Tn <- sub(".*gene=", "", transposons$tag)
transposons$ID <- paste(transposons$assembly, transposons$Tn, sep = ":")

#Correct for TnUncertain
SRR1520067_toKepp <- transposons[which(transposons$assembly=="SRR1520067" & 
                                         transposons$end - transposons$start == max( transposons[assembly=="SRR1520067"]$end -  transposons[assembly=="SRR1520067"]$start)),]
transposons <- rbind(transposons[assembly != "SRR1520067",], #Remove all lines with the assembly
                     #Add the one we chose to keep
                     #And modify name of its Tn
                     mutate(SRR1520067_toKepp, Tn = "TnUncertain", ID = "SRR1520067:TnUncertain"))

#Save this for further use
fwrite(transposons, "transposons100.tsv", sep = '\t')


transposons$loc = paste(transposons$assembly, transposons$contig, sep = ":")

#Tell which ones in plasmids
transposons[, molecule_type:=ifelse(loc %in% plasmids$loc, "plasmid", "chromosome")]


#Create the empty matrix
tnMatrix <- matrix(nrow = 936, ncol = length(unique(transposons$Tn)))
rownames(tnMatrix) <- tree$tip.label
colnames(tnMatrix) <- unique(transposons$Tn)


#Fill the matrix
for(j in colnames(tnMatrix)){ 
  for(i in rownames(tnMatrix)){ 
    print(i)
    toFill <-  transposons[assembly == i & Tn==j,]
    
    if(nrow(toFill)>0){
      if(length(unique(toFill$molecule_type))==1 & unique(toFill$molecule_type)=="plasmid"){
        tnMatrix[i,j] =  2
      } else if(length(unique(toFill$molecule_type))==1 & unique(toFill$molecule_type)=="chromosome"){
        tnMatrix[i,j] =  1
      } else if(length(unique(toFill$molecule_type))==2){
        tnMatrix[i,j] =  3
      } 
    } else {
      tnMatrix[i,j] =  0
    }
  } 
} 

#Also give color to transposon
tnMatrix <- tnMatrix * 10

###########

### 3rd matrix = PHAGES

#For this, we need this data:
phages <- allMGE[type=="phage"] %>%
  mutate(., contig = sub("[||].*", "", splitToColumns(MGEname, split = ":", columns = 2)))  %>%
  mutate(., loc = paste(assembly, contig, sep = ":"))

#Tell which ones in plasmids
phages[, molecule_type:=ifelse(loc %in% plasmids$loc, "plasmid", "chromosome")]


#Create the empty matrix
phageMatrix <- matrix(nrow = 936, ncol = 1)
rownames(phageMatrix) <- tree$tip.label
colnames(phageMatrix) <- "phage"

#Fill the matrix
for(i in rownames(phageMatrix)){ 
  print(i)
  toFill <-  phages[assembly == i,]
    
  if(nrow(toFill)==0){
    phageMatrix[i,1] =  0
  } else {
    molType = unique(toFill$molecule_type)
    if(length(molType)==2){
      phageMatrix[i,1] =  3
    } else {
      if(molType=="chromosome"){
        phageMatrix[i,1] =  1 
      } else {
        phageMatrix[i,1] =  2
      }
    }
  }
} 

phageMatrix = phageMatrix*100

###########

### 4th matrix = genes

genesInterest <- distinct(annotCare[Element_type.amrfinder == "STRESS" | Element_type.amrfinder == "AMR", c("geneUndup", "Element_type.amrfinder")])
setorder(genesInterest, Element_type.amrfinder, geneUndup)
genesInterest[, colType:=ifelse(Element_type.amrfinder == "AMR", "forestgreen", "orange")]


#Create a matrix with one row per assembly & one column per gene
annotMatrix <- matrix(nrow = 936, ncol = nrow(genesInterest))
rownames(annotMatrix) <- tree$tip.label
colnames(annotMatrix) <- genesInterest$geneUndup


#The matrix about genes
for(j in colnames(annotMatrix)){ 
  for(i in rownames(annotMatrix)){ 
    print(i)
    toFill <- annotCare[assembly == i & geneUndup == j & (Element_type.amrfinder=="AMR" | Element_type.amrfinder == "STRESS"),] 
    
    if(nrow(toFill)==0){
      annotMatrix[i,j] =  0
    } else {
      
      #alpha says whether on chromosome (1) or plasmid (2) or depend (3)
      if(length(unique(toFill$molecule_type))>1){
        alpha <- 3 #depends on copy
      } else if(length(unique(toFill$molecule_type))==1 & unique(toFill$molecule_type)=="chromosome"){
        alpha <- 1
      } else if(length(unique(toFill$molecule_type))==1 & unique(toFill$molecule_type)=="plasmid") {
        alpha <- 2
      } else if(length(unique(toFill$molecule_type))==1 & unique(toFill$molecule_type)=="PP") {
        alpha <- 4
      }
      
      #No matter alpha, gene can be in MGE (transposon or phage) or not
      if(length(unique(toFill$MGE))>1){
        MGE <- 1000 #depends on copy
      } else if(length(unique(toFill$MGE))==1 & unique(toFill$MGE)=="transposon"){
        MGE <- 10
      } else if(length(unique(toFill$MGE))==1 & unique(toFill$MGE)=="phage") {
        MGE <- 100
      }  else if(length(unique(toFill$MGE))==1 & unique(toFill$MGE)=="") {
        MGE <- 1 #No MGE
      }
      
      #Each combination of molType & MGE, get its own number
      annotMatrix[i,j] = alpha*MGE
    }
  } 
}  
  

############

#Bind all four matrix + leave add an empty matrux after annotMatrix to generate a space in the figure
matrixBind <- cbind(annotMatrix, matrix(nrow = 936, ncol = 1), tnMatrix, plasmidMatrix, phageMatrix)

#Order assemblies in same order as in tree
matrixBind <- matrixBind[match(tree$tip.label, rownames(matrixBind)),]

#SAve this marix because super long to generate
saveRDS(matrixBind, file = "matrixAnnotation_MGE_and_genes.RData") 


###################################
###### Investigate virulence ######
#### And then add it in matrix ####
###################################

# Identify hypo and huer virulence
#Based on Maury et al. 2016

#### truncated inlA associated with hypovurulance

#NCBI inlA EDG-g is 2402bp long
2402-(3*20) #Maury considers smaller of than 20aa

annot[geneUndup=="inlA" & end-start>2300 & end-start<2350, end-start] #Not a single gene in beween 2300 and 2350bp so would be good to put threashold there

hypo <- annot[geneUndup=="inlA"] %>% mutate(., inlA = ifelse(end-start<2300, "truncated", "full"))
hypo_summarize <- hypo %>% group_by(assembly) %>% summarize(nVir = length(unique(inlA)), inlA = ifelse(length(unique(inlA))==1, inlA, "depends")) %>% data.table()
hypo_summarize[nVir==2] #101 assemblies have a full gene + a truncated one
hypo_summarize[inlA=='truncated'] #Only 6 assemblies always have it truncated
hypo_summarize[inlA=='full'] #829 never have it truncated

#Number of genomes with 1 to 4 copies of inlA
table(sort(table(hypo$assembly))) 

### Hyper virulence : LIPI
#We annoated those with the Institut Pasteur BIGSdb Listeria database 

LIPI <- data.table(read_excel("LIPI_annotation.xlsx", skip=2))

which(str_detect(colnames(LIPI), "LIPI2"))
LIPI2 <- LIPI[1:936, 64:75]
which(str_detect(colnames(LIPI), "LIPI3"))
LIPI3 <- LIPI[1:936, 77:84]
which(str_detect(colnames(LIPI), "LM9"))
LIPI4 <- LIPI[1:936, 86:91]

LIPI1 <-  LIPI[1:936, c("prfA", "plcA", "hly", "mpl", "actA", "plcB")]

LIPI_summary <- data.table(assembly = LIPI[1:936,]$ID,  LIPI1 = rowSums(LIPI1), LIPI2 = rowSums(LIPI2), LIPI3 = rowSums(LIPI3), LIPI4 = rowSums(LIPI4))

min(LIPI_summary$LIPI1) #They all have LIPI1 so tot interesting
table(LIPI_summary$LIPI2) #Isolate never have more than 1 gene of LIPI2 so they don't really have it

# --> Focus only on inlA, LIPI3 and LIPI4
virulence = left_join(hypo_summarize[, -"nVir"], LIPI_summary[, c("assembly", "LIPI3","LIPI4")])

#Make a code so can be plot at the same time as the other matrix
virulence[, LIPI3:=ifelse(LIPI3==0, 0, ifelse(LIPI3==8, 5000, 4000))]
virulence[, LIPI4:=ifelse(LIPI4==0, 0, ifelse(LIPI4==6, 5000, 4000))]
virulence[, inlA:=ifelse(inlA %in% c("depends", 'full'), 1, ifelse(inlA=="truncated", 6000, NA))]

virulence <- virulence[match(tree$tip.label, assembly),]

matrixVirulence = as.matrix(virulence[, -'assembly'])
rownames(matrixVirulence) = virulence$assembly

saveRDS(matrixVirulence, file = "matrixVirulence.RData") 




##################################
######## Plot the matrix ########
###### Figure 1 of the paper #####
##################################

matrixBind_all <- cbind(matrixVirulence, matrix(nrow = 936, ncol = 1), matrixBind)

#Those are all the possible values in my matrix:
unique(as.vector(matrixBind_all))

#Note: because it is not possible to have two RowSide, I generated the exact same figure twice, one with CC in RowSide, the other with lineages in RowSide
#Then, I put thel together in inskape
#To generate the figure with lineage, uncomment lines related to lineage, and comment those related to CC

pdf("Figure1_CC.pdf", width = 15, height = 10)
#pdf("Figure1_Lineage.pdf", width = 15, height = 10)

heatmap(matrixBind_all, Rowv=dend, Colv=NA, margins=c(5.5,0), scale="none",
        #too many col, need to write smaller
        cexCol = 1,
        labRow = NA,
        #RowSideColors = c(colLineage$colLineage),
        RowSideColors = colLineage$colCC2,
        ColSideColors = c("lightslateblue", "maroon", "maroon", "white", genesInterest$colType,
                          "white",
                          rep("blue", 6), #6 Tn
                          rep("gray47", 21), #21 Pl
                          rep("deeppink", 1) #Just 1 column for phage
        ),
        #Give breaks by hand: eg a same color for values from 0 to 0.9, another from 0.9 to 1.1, etc
        #ATTENTION normally we would also need 200.1, but here we have no 200 (no phage in plasmid)
        breaks =c(0, 0.9, 1.1, 2.1, 3.1, 4.1, 10.1, 20.1, 100.1, 300.1, 3000.1, 4000.1, 5000.1, 6000.1), 
        col = c("white", "gray89", "gray47", "black", "orange", "cyan", "blue","deeppink",  "black","black","orangered", "red3","lightslateblue")
)                
                
              
legend("bottomleft", 
       #how far from the bottomleft
       inset = c(0.005,0.82),
       lwd = 4, #size col
       cex = 0.8, #size text
       legend = paste("Lineage", 
                      rev(distinct(colLineage[, c("phylogenetic_lineage", "colLineage")])$phylogenetic_lineage),
                      sep = " "),
       col = rev(distinct(colLineage[, c("phylogenetic_lineage", "colLineage")])$colLineage),
       title = "Row 1"
)             

legend("bottomleft", 
       #how far from the bottomleft
       inset = c(0.1,0.76), 
       lwd = 4, #size col
       cex = 0.8, #size text
       legend =  distinct(colLineage_legend[!colCC2 %in% grays, c("CC", "colCC2")])$CC,
       col = distinct(colLineage_legend[!colCC2 %in% grays, c("CC", "colCC2")])$colCC2,
       title = "Row 2"
)

#Add legend for type genes 
legend("bottomleft", 
       #inset = c(0.06,0.58),
       inset = c(0.04,0.55),
       lwd = 4, #size col
       cex = 0.8, #size text
       legend = c("Hypovirulence", "Hypervirulence",
                  "AMR genes",
                  "Stress genes",
                  "Transposons",
                  "Plasmids",
                  "Phages"),
       col = c("lightslateblue", "maroon",
               distinct(genesInterest[, c("Element_type.amrfinder", "colType")])$colType,
               "blue",
               "gray47",
               "deeppink"),
       title = "Top colors"
)

legend("bottomleft", 
       inset = c(0.55,0.6),
       lwd = 4, #size col
       cex = 0.8, #size text
       legend = c("chromosome (no MGE)", "plasmid", "PP element", "chromosomic transposon", "plasmidic transposon", "phage",  "multiple"),
       col =c("gray89", "gray47", "orange", "cyan", "blue", "deeppink", "black"),
       title = "Vehicles"
)

legend("bottomleft", 
       inset = c(0.037,0.4),
       lwd = 4, #size col
       cex = 0.8, #size text
       legend = c("full gene", "truncated gene", ">0 gene of LIPI", "all genes of LIPI"),
       col = c("gray89", "lightslateblue", "orange", "red"),
       title = "Virulence"
)

dev.off()   



##################################
## With same data as for matrix, ##
# generate Figure 4 of the paper #
##################################

###############
### PANEL A ###
###############
annotCare[molecule_type=="chromosome" & MGE=="phage",]$molecule_type = "phage"

#I actually modify line 150 so phage is separted from chrm
summarizePercGenes_where <- annotCare[Element_type.amrfinder=="AMR",] %>%
  group_by(molecule_type, MGE) %>%
  #Count the number of CDS (here core)
  summarize(nCDS=n()) %>% 
  mutate(., 
         kind = "AMR",
         perc = nCDS*100/nrow(annotCare[Element_type.amrfinder=="AMR"])) %>%
  #To which we add info about CDS plus
  rbind(., 
        annotCare[Element_type.amrfinder=="STRESS",] %>% 
          group_by(molecule_type, MGE) %>% 
          summarize(nCDS=n()) %>% 
          mutate(.,
                 kind = "Stress",
                 perc = nCDS*100/nrow(annotCare[Element_type.amrfinder=="STRESS"]))
  )  %>%
  #And then info about CDS others
  #I choose to not include genes whch names are also in AMR or plus
  rbind(., 
        #annotCare[kind=="CDS others" & !(geneUndup %in% annotCare[scope=="CDS plus" | scope=="CDS core",]$geneUndup),] %>% 
        #We can also see if same results when remove hypothetcal proteins (don't forget to also uncomment perc=)
        annotCare[type.prokka=="CDS" & product.prokka!="hypothetical protein" & !(geneUndup %in% annotCare[kind!="",]$geneUndup),] %>%  
          group_by(molecule_type, MGE) %>% 
          summarize(nCDS=n()) %>% 
          mutate(.,
                 kind = "others",
                 #perc = nCDS*100/nrow(annotCare[scope=="CDS others" &  !(geneUndup %in% annotCare[scope=="CDS plus" | scope=="CDS core",]$geneUndup),]))
                 perc = nCDS*100/nrow(annotCare[type.prokka=="CDS" & product.prokka!="hypothetical protein" & !(geneUndup %in% annotCare[kind!="",]$geneUndup),]))
  )  %>%
  data.table

summarizePercGenes_where[, titleFig:=ifelse(kind=="AMR", "CDS AMR (NS)", ifelse(kind=="Stress", "CDS Stress (***)", "CDS others"))]

p1 <- ggplot(summarizePercGenes_where[molecule_type!="PP"], aes(x = molecule_type, y = perc, fill = paste(molecule_type, MGE, sep = ":"))) +
  geom_bar(stat = "identity") +
  facet_wrap(~titleFig) +
  theme_bw() +
  ylim(0,100) +
  xlab("") +
  ylab("Percentage of CDS in each vehicle") +
  scale_fill_manual(values = c("chromosome:"="gray89",
                               "chromosome:transposon"="cyan",
                               "phage:phage"="deeppink",
                               "plasmid:"="gray47",
                               "plasmid:phage"="red",
                               "plasmid:transposon"="blue"),
                    labels = c("in chromosome (no MGE)", 
                               "in chromosomic transposon",
                               "in phage",
                               "in plasmid", 
                               "in plasmidic phage", 
                               "in plasmidic transposon")) +
  guides(fill=guide_legend(title="Gene vehicles"))

#Test stat comparing all MGE vs no MGE
chi2_noMGE <- summarizePercGenes_where[molecule_type=="chromosome" & MGE==""] %>% group_by(kind) %>% summarize(nCDS_tot_noMGE = sum(nCDS))
chi2_MGE <- summarizePercGenes_where[molecule_type!="PP" & (molecule_type!="chromosome" | MGE!="")] %>% group_by(kind) %>% summarize(nCDS_tot_MGE = sum(nCDS))
chi2_bis <-  left_join(chi2_noMGE, chi2_MGE, by = "kind")  %>% data.table()
chi2_bis <-  t(chi2_bis[, -1])
colnames(chi2_bis) <- c("AMR", "STRESS", "others")
fisher.test(chi2_bis[, c("AMR", "others")], simulate.p.value = TRUE )
fisher.test(chi2_bis[, c("STRESS", "others")], simulate.p.value = TRUE )

###############
### PANEL B ###
###############
#0 plasmids have at least 1 AMR gene
plasmids_withAMR <- data.table()
  
plasmids_withStress <- annotCare[molecule_type=="plasmid" & Element_type.amrfinder=="STRESS", ] %>% 
  left_join(., plasmids[, c("loc", "primary_cluster_id")], by = "loc") %>%
  select(., c("primary_cluster_id", "assembly")) %>%
  distinct() %>%
  mutate(., ID_Pl = paste(primary_cluster_id, assembly, sep = ":"))

#And then transposons
transposon_withAMR <- annotCare[MGE=="transposon" & Element_type.amrfinder=="AMR", c("geneUndup", "assembly", "contig", "inTransposon")] %>% 
  distinct() %>% 
  mutate(., ID_Tn = paste(splitToColumns(inTransposon,split = ":", columns = 2 ), assembly, contig, sep = ":"))

transposon_withStress <- annotCare[MGE=="transposon" & Element_type.amrfinder=="STRESS", c("geneUndup", "assembly", "contig", "inTransposon")] %>% 
  distinct() %>% 
  mutate(., ID_Tn = paste(splitToColumns(inTransposon,split = ":", columns = 2 ), assembly, contig, sep = ":"))

transposon_withBoth <- transposon_withStress[ID_Tn %in% transposon_withAMR$ID_Tn] #0



percGene_inMGE <- data.frame(type = c("plasmid",  "transposon"), 
                             nTot = c(length(unique(plasmids$nodeName)), nrow(transposons)),
                             n_withAMR = c(nrow(plasmids_withAMR), nrow(transposon_withAMR)),
                             n_withStress = c(nrow(plasmids_withStress),nrow(transposon_withStress))
) %>%
  mutate(., 
         perc_withAMR = n_withAMR*100/nTot,
         perc_withStress = n_withStress*100/nTot)

#Write data in such a way I can plot them
percGene_inMGE_fig <- data.table(type = rep(percGene_inMGE$type, time = 2),
                                 perc = c(percGene_inMGE$perc_withAMR, percGene_inMGE$perc_withStress),
                                 gene = c(rep(c("AMR", "Stress"), each =2))
)

p2 <- ggplot(percGene_inMGE_fig, aes(x = type, y = perc, fill = gene)) +
  geom_bar(stat = "identity") +
  theme_bw() +
  ylim(0,100) +
  xlab("") +
  ylab("Percentage of MGE with such gene") +
  scale_fill_manual(values = c("AMR" = "green", "Stress"= "orange")) +
  guides(fill=guide_legend(title="Gene type"))

transposon_withStress <-mutate(transposon_withStress, loc=paste(assembly, contig, sep = ":")) %>% left_join(., plasmids[, c("primary_cluster_id", "loc")])
nrow(transposon_withStress[!is.na(primary_cluster_id)])*100/nrow(transposon_withStress)

###############
### PANEL C ###
###############
#Instead of looking just exchanges between chromsomome and plasmid, look at exchanges with any vehicles
geneExchanges_anyV <- distinct(annotCare[(type.prokka=="CDS" & product.prokka!="hypothetical protein") | Element_type.amrfinder!="",
                                         c("geneUndup", "Element_type.amrfinder", "molecule_type", "MGE", "assembly")]) %>% 
  filter(., geneUndup!="") %>%
  filter(., molecule_type!="PP") %>%
  filter(., (!(geneUndup %in% annotCare[Element_type.amrfinder!=""]$geneUndup & Element_type.amrfinder=="")) | Element_type.amrfinder!="") %>%
  mutate(., where = paste(molecule_type, MGE, sep = ":")) %>% 
  group_by(geneUndup, Element_type.amrfinder) %>%
  dplyr::summarize(nbVehicles = length(unique(where))) %>%
  mutate(., titleFig = ifelse(Element_type.amrfinder=="STRESS", "Gene Stress (***)", 
                ifelse(Element_type.amrfinder=="AMR", "Gene AMR (NS)", "Gene others"))) %>%
  data.table() 

p3 <- ggplot(geneExchanges_anyV,  aes(x=nbVehicles)) + geom_bar(stat = "count", aes(fill=titleFig)) +
  facet_wrap(~titleFig, scales="free_y") +
  theme_bw() +
  xlab("Number of vehicle types in which the gene is found") +
  ylab("Number of genes") +
  scale_fill_manual(values = c("Gene AMR (NS)" = "green", "Gene Stress (***)"= "orange", "Gene others" = "darkgray")) +
  theme(legend.position = "none")

#Statistical test
geneExchanges_anyV_stat <- full_join(geneExchanges_anyV[titleFig=="Gene others"] %>% group_by(nbVehicles) %>% summarize(nGenes.others = n()),
                                     geneExchanges_anyV[titleFig=="Gene Stress (***)"] %>% group_by(nbVehicles) %>% summarize(nGenes.stress = n()),
                                     by = "nbVehicles") %>% 
  full_join(.,  geneExchanges_anyV[titleFig=="Gene AMR (NS)"] %>% group_by(nbVehicles) %>% summarize(nGenes.AMR = n())) %>%
  data.table()
geneExchanges_anyV_stat[is.na(nGenes.stress)]$nGenes.stress = 0
geneExchanges_anyV_stat[is.na(nGenes.AMR)]$nGenes.AMR = 0

fisher.test(geneExchanges_anyV_stat[, c("nGenes.others", "nGenes.stress")])
fisher.test(geneExchanges_anyV_stat[, c("nGenes.others", "nGenes.AMR")])

###############
### Put all pannels of figure 4 together
###############

gt <- arrangeGrob(p1, p2, p3, 
                  layout_matrix = rbind(c(1,1,1,1,1), c(2,2,3,3,3)))

#pdf("Figure3.pdf", width = 10, height = 8)

as_ggplot(gt) +                                # transform to a ggplot
  draw_plot_label(label = c("A", "B", "C"), size = 15,
                  x = c(0, 0, 0.4), y = c(1, 0.5, 0.5))
dev.off()


